lib/Doctrine/ORM/AbstractQuery.php

```diff
@@ -859,6 +859,31 @@ public function iterate($parameters = null, $hydrationMode = null)
         return $this->em->newHydrator($this->hydrationMode)->iterate($stmt, $rsm, $this->hints);
     }
 
+    /**
+     * Executes the query and returns an iterable that can be used to incrementally
+     * iterate over the result.
+     *
+     * @param ArrayCollection|mixed[]|null $parameters    The query parameters.
+     * @param string|int|null              $hydrationMode The hydration mode to use.
+     *
+     * @return iterable<mixed>
+     */
+    public function getIterable(?iterable $parameters = null, $hydrationMode = null) : iterable
+    {
+        if ($hydrationMode !== null) {
+            $this->setHydrationMode($hydrationMode);
+        }
+
+        if ($parameters !== null && count($parameters) !== 0) {
```

Comment:
I'm not sure about using count here for an `iterable`. The doc comment reduces the type to `ArrayCollection` and `array`, but technically every iterable can be injected. Should an extra complexity be added with `iterator_count()`?

Reply:
AFAIK `iterator_count` takes only `\Iterator` and not `iterable`. 

But you're correct that it's technically possible to pass every `iterable` and then it would crash. However, removing that type hint would have same effect, right? It would allow to pass even more types and then we would have to if-check for proper type anyway. So I still think the proper parameter type is defined in PHPDoc and `iterable` typehint is just kind of weak guarantee.

WDYT? I don't have a preferred way to handle it but would like to avoid something like `if not instance of arraycollection and is not array and is not null then throw`

Reply:
`?iterable` is fine, an improvement and omitting the type leads to the mentioned type misbehaviour, which is why I propose to add complexity and use both counts. `Traversable` can't be used alone without `Iterator` and `IteratorAggregate` for an instantiable custom class and even `IteratorAggregate` can be used in `iterator_count()` as seen in this example: https://3v4l.org/8j5mf

`count()` on an iterable instance implementing `Countable` should be preferred before `iterator_count()` to keep more BC to the previous ORM versions, if there's no argument against this suggestion.

Disadvantage: `setParameters()` itself, which is called there, checks if the set parameters variable is an array. In case of an `array`, it creates an `ArrayCollection` instance and set the elements of the array in there. In the current implementation of `setParameters()` it would set any other iterable to `$this->parameters`. Sure, it also did it in the versions before as long as the parameters were countable. Does `setParameters()` need to be adapted too? This PR would become bigger if that method is also going to be adapted. :thinking:

Reply:
I tend to keep `setParameters()` as it is for now and focus only on `getIterable()`.

Reply:
@SenseException is this what you meant?

```php
        if ($parameters !== null
            && (
                (($parameters instanceof Countable || is_array($parameters)) && count($parameters) !== 0)
                || ($parameters instanceof Traversable && iterator_count($parameters) !==0)
            )
        ) {
```

Reply:
``` php
        if ($parameters !== null
            && (
                (($parameters instanceof Countable || is_array($parameters)) && count($parameters) !== 0)
                || iterator_count($parameters) !== 0
            )
        ) {
```

Too bad we don't have `is_countable()` in the lowest supported PHP version, but I guess you could move `is_array()` and the `Countable` check into a private method. Because of the `?iterable` type hint, all that is left are iterators after the array check. What do you think about this?